waspc/src/Wasp/Generator/DbGenerator.hs

```diff
@@ -83,42 +83,67 @@ genMigrationsDir spec =
 --     Common scenarios for the second warning include:
---       - After a fresh checkout, or after `wasp clean`; possible false positives in these cases, but for safety, it's still preferable to warn.
+--       - After a fresh checkout, or after `wasp clean`.
 --       - When they previously had no entities and just added their first.
-warnIfDbSchemaChangedSinceLastMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO (Maybe GeneratorWarning)
-warnIfDbSchemaChangedSinceLastMigration spec projectRootDir = do
+--     In either of those scenarios, validate against DB itself to avoid redundant warnings.
+--
+--     NOTE: As one final optimization, if they do not have a schema.prisma.wasp-last-db-concurrence-checksum but the schema is
+--     in sync with the databse, we generate that file to avoid future checks.
+warnIfDbNeedsMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO (Maybe GeneratorWarning)
+warnIfDbNeedsMigration spec projectRootDir = do
   dbSchemaChecksumFileExists <- doesFileExist dbSchemaChecksumFp
-
   if dbSchemaChecksumFileExists
-    then do
-      dbSchemaFileChecksum <- hexToString <$> checksumFromFilePath dbSchemaFp
-      dbChecksumFileContents <- readFile dbSchemaChecksumFp
-      return $ warnIf (dbSchemaFileChecksum /= dbChecksumFileContents) "Your Prisma schema has changed, you should run `wasp db migrate-dev`."
-    else return $ warnIf entitiesExist "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."
+    then warnIfSchemaDiffersFromChecksum dbSchemaFp dbSchemaChecksumFp
+    else
+      if entitiesExist
+        then warnIfSchemaDiffersFromDb projectRootDir
+        else return Nothing
   where
     dbSchemaFp = SP.fromAbsFile $ projectRootDir </> dbSchemaFileInProjectRootDir
-    dbSchemaChecksumFp = SP.fromAbsFile $ projectRootDir </> dbSchemaChecksumOnLastMigrateFileProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ projectRootDir </> dbSchemaChecksumOnLastDbConcurrenceFileProjectRootDir
     entitiesExist = not . null $ getEntities spec
 
-    warnIf :: Bool -> String -> Maybe GeneratorWarning
-    warnIf b msg = if b then Just $ GeneratorNeedsMigrationWarning msg else Nothing
+warnIfSchemaDiffersFromChecksum :: FilePath -> FilePath -> IO (Maybe GeneratorWarning)
+warnIfSchemaDiffersFromChecksum dbSchemaFp dbSchemaChecksumFp = do
+  dbSchemaFileChecksum <- hexToString <$> checksumFromFilePath dbSchemaFp
+  dbChecksumFileContents <- readFile dbSchemaChecksumFp
+  if dbSchemaFileChecksum /= dbChecksumFileContents
+    then return . Just $ GeneratorNeedsMigrationWarning "Your Prisma schema has changed, please run `wasp db migrate-dev` when ready."
+    else return Nothing
+
+warnIfSchemaDiffersFromDb :: Path' Abs (Dir ProjectRootDir) -> IO (Maybe GeneratorWarning)
+warnIfSchemaDiffersFromDb projectRootDir = do
+  -- NOTE: If we wanted to, we could also check that the migrations dir == db,
+  -- but a schema check should handle all most likely cases.
```

Comment:
Aha I think this what I was asking in other comments in this PR!
So how is that we don't check for migrations dir? Isn't that also important? What can happen if we don't check it, how could that be a problem, what are the scenarios? I know we talked about but I forgot it could you remind me a bit? Maybe we should also document that just a bit more hm.

Reply:
I will add a deeper comment on the rationale for now vs addressing all the migrate questions in the comments. If we don't check it, the worst case is we don't warn them to migrate. :) So not catastrophic, but not ideal if we miss anything.

Reply:
ok great, deeper comment is what I think will solve this!
I commented on other place on this also. I will now resolve all the issues asking for this same thing sorry about that.